File: src/lib/game/loop.ts
import { tick, getState } from './state.js'
import { broadcastToMatch } from '../websocket/server.js'
import { getAllMatches } from '../match/manager.js'
import type { StateMessage } from '../../types.js'

const TICK_RATE = 30 // Hz
const TICK_INTERVAL = 1000 / TICK_RATE

let loopInterval: ReturnType<typeof setInterval> | null = null

export function startGameLoop(): void {
  if (loopInterval) return

  console.log(`[GAME] Starting game loop at ${TICK_RATE} Hz`)

  loopInterval = setInterval(() => {
    // Update all active matches
    for (const match of getAllMatches()) {
      tick(match.matchId)

      const state = getState(match.matchId)
      if (state) {
        const message: StateMessage = {
          type: 'STATE',
          state,
        }
        broadcastToMatch(match.matchId, message)
      }
    }
  }, TICK_INTERVAL)
}

export function stopGameLoop(): void {
  if (loopInterval) {
    clearInterval(loopInterval)
    loopInterval = null
    console.log('[GAME] Game loop stopped')
  }
}

File: internal/game/loop.go
package game

import (
	"fmt"
	"time"

	"space-invaders-coop/backend-go/internal/match"
	"space-invaders-coop/backend-go/internal/types"
)

const tickRate = 30 // Hz
const tickInterval = time.Second / tickRate

// Broadcaster is used to send state to clients (e.g. WebSocket hub).
type Broadcaster interface {
	BroadcastToMatch(matchID string, msg interface{})
}

var (
	loopRunning bool
	stopCh      chan struct{}
)

// StartLoop starts the game loop. It ticks all matches and broadcasts state via b.
func StartLoop(b Broadcaster) {
	if loopRunning {
		return
	}
	loopRunning = true
	stopCh = make(chan struct{})
	fmt.Printf("[GAME] Starting game loop at %d Hz\n", tickRate)
	ticker := time.NewTicker(tickInterval)
	defer ticker.Stop()
	for {
		select {
		case <-stopCh:
			loopRunning = false
			fmt.Println("[GAME] Game loop stopped")
			return
		case <-ticker.C:
			tickStart := time.Now()
			var broadcastDur time.Duration
			for _, m := range match.GetAllMatches() {
				Tick(m.MatchID)
				state := GetState(m.MatchID)
				if state != nil {
					bcStart := time.Now()
					b.BroadcastToMatch(m.MatchID, types.StateMessage{Type: "STATE", State: *state})
					broadcastDur += time.Since(bcStart)
					if state.GameOver && match.MarkPersisted(m.MatchID) {
						outcome := "defeat"
						if state.Victory {
							outcome = "victory"
						}
						if snap := match.Snapshot(m.MatchID, outcome); snap != nil {
							match.FireFinalizer(snap)
						}
					}
				}
			}
			RecordTick(time.Since(tickStart))
			RecordBroadcast(broadcastDur)
		}
	}
}

// StopLoop stops the game loop.
func StopLoop() {
	if stopCh != nil {
		close(stopCh)
	}
}
